pySDC/tests/test_projects/test_pintsime/test_SwitchEstimator.py

```diff
@@ -0,0 +1,464 @@
+        t0=t0,
+        Tend=Tend,
+        exact_event_time_avail=exact_event_time_avail,
+    )
+
+    sum_restarts = np.sum(np.array(get_sorted(stats, type='restart', sortby='time', recomputed=None))[:, 1])
+    assert sum_restarts == 0, 'Event occurs at boundary, but restart(s) are executed anyway!'
+
+
+@pytest.mark.base
+@pytest.mark.parametrize('tol', [10 ** (-m) for m in range(8, 13)])
+@pytest.mark.parametrize('num_nodes', [3, 4, 5])
+@pytest.mark.parametrize('quad_type', ['LOBATTO', 'RADAU-RIGHT'])
+def testDetectionODE(tol, num_nodes, quad_type):
+    r"""
+    Here, the switch estimator is applied to a dummy problem of ``DiscontinuousTestODE``,
+    where the dynamics of the differential equation is replaced by its exact dynamics to see if
+    the switch estimator predicts the event correctly. The problem is tested for a combination
+    of different tolerances ``tol`` and different number of collocation nodes ``num_nodes``.
+
+    Since the problem only uses the exact dynamics, the event should be predicted very accurately
+    by the switch estimator.
+
+    Parameters
+    ----------
+    tol : float
+        Tolerance for switch estimator.
+    num_nodes : int
+        Number of collocation nodes.
+    quad_type : str
+        Type of quadrature.
+    """
+
+    from pySDC.implementations.sweeper_classes.generic_implicit import generic_implicit
+    from pySDC.helpers.stats_helper import get_sorted
+    from pySDC.projects.PinTSimE.battery_model import generateDescription, controllerRun
+    from pySDC.implementations.hooks.log_solution import LogSolution
+    from pySDC.implementations.hooks.log_restarts import LogRestarts
+
+    problem = ExactDiscontinuousTestODE
+    problem_params = dict()
+    t0 = 1.6
+    Tend = 1.62
+    dt = Tend - t0
+
+    sweeper = generic_implicit
+    QI = 'IE'
+
+    restol, alpha, maxiter, max_restarts, useA, useSE, exact_event_time_avail = getParamsRun()
+
+    hook_class = [LogSolution, LogRestarts]
+
+    description, controller_params, controller = generateDescription(
+        dt=dt,
+        problem=problem,
+        sweeper=sweeper,
+        num_nodes=num_nodes,
+        quad_type=quad_type,
+        QI=QI,
+        hook_class=hook_class,
+        use_adaptivity=useA,
+        use_switch_estimator=useSE,
+        problem_params=problem_params,
+        restol=restol,
+        maxiter=maxiter,
+        max_restarts=max_restarts,
+        tol_event=tol,
+        alpha=alpha,
+    )
+
+    stats, t_switch_exact = controllerRun(
+        description=description,
+        controller_params=controller_params,
+        controller=controller,
+        t0=t0,
+        Tend=Tend,
+        exact_event_time_avail=exact_event_time_avail,
+    )
+
+    # in this specific example only one event has to be found
+    switches = [me[1] for me in get_sorted(stats, type='switch', sortby='time', recomputed=False)]
+    assert len(switches) >= 1, f'{problem.__name__}: No events found for tol={tol}!'
```

Comment:
In the Jülich chat someone recently posted that you can do string formatting like `{tol=}` and it will show as "toll=3.14" or whatever. Of course, you don't need to change this minor detail. But I use it all the time now and want to pass this on :)

Reply:
Ok, I'll try that out!